doc/man/man8/ldif2ldbm.8
.TH LDIF2LDBM 8C "22 September 1998" "OpenLDAP LDVERSION"
.SH NAME
ldif2ldbm, ldif2index, ldif2id2entry, ldif2id2children \- LDIF to LDBM database format conversion utilities
.SH SYNOPSIS
.B SBINDIR/ldif2ldbm
.B \-i ldif\-input\-file
.B [\-d debug\-level] [\-f slapd\-config\-file]
.B [\-j number\-of\-jobs] [\-n database#] [\-s sbindir]
.LP
.B SBINDIR/ldif2index
.B \-i ldif\-input\-file
.B [\-d debug\-level] [\-f slapd\-config\-file] [\-n database#]
.B attribute\-name
.LP
.B SBINDIR/ldif2id2entry
.B \-i ldif\-input\-file
.B [\-d debug\-level] [\-f slapd\-config\-file] [\-n database#]
.LP
.B SBINDIR/ldif2id2children
.B \-i ldif\-input\-file
.B [\-d debug\-level] [\-f slapd\-config\-file] [\-n database#]
.LP
.SH DESCRIPTION
.LP
These programs are used to convert a database in text LDIF LDAP
Directory Interchange Format (LDIF) to an LDBM database suitable
for use by
.BR slapd (8).
Normally, only
.B ldif2ldbm
is invoked by you. It will invoke the other programs as necessary.
Occasionally, it may be necessary to invoke them by hand. For
example, to create a new index file for an existing database, the
.B ldif2index
program can be invoked. The
.BR ldbmcat (8)
program is used to do the reverse conversion.
.LP
See "The SLAPD and SLURPD Administrator's Guide" for more details on
using these programs.
.SH OPTIONS
The first three options apply to all four programs. The -j option is
only for the 
.B ldif2ldbm
program.
.TP
.BI \-i " ldif\-input\-file"
This option specifies the location of the LDIF input file containing
the database to convert. It is required.
.TP
.BI \-d " debug\-level"
Turn on debugging as defined by
.B debug\-level.
Some general operation and status messages are printed for any value of
\fIdebug\-level\fP.  \fIdebug\-level\fP is taken as a bit string, with
each bit corresponding to a different kind of debugging information. 
See <ldap.h> for details.
.TP
.BI \-f " slapd\-config\-file"
This option specifies the
.B slapd
configuration file. The default is
.BR ETCDIR/slapd.conf .
.TP
.BI \-n " database#"
This option specifies the database number within
.B slapd
configuration file to build. The default is the 1, the first.
.TP
.BI \-s " sbindir"
This option specifies the location of dependent tools.  The
default is SBINDIR.
.TP
.BI \-j " number\-of\-jobs"
This option only applies to the
.B ldif2ldbm
program. It specifies the level of parallelism to use when doing the
conversion.
.B ldif2ldbm
invokes several other programs during the conversion process,
most notably one invocation of
.B ldif2index
for each indexed attribute that appears in the LDIF input file. The -j
option tells
.B ldif2ldbm
how many of these other programs it should run in parallel. This can
speed up the conversion, but beware of starting too many processes
in parallel, all competing for disk, memory, and cpu resources. The
default is one.
.SH EXAMPLES
To convert the file
.BR ldif.input
into an LDBM database with indexes as described in the
.I slapd
config file
.BR ETCDIR/slapd.conf ,
give the command:
.LP
.nf
.ft tt
	SBINDIR/ldif2index -i ldif.input -f ETCDIR/slapd.conf
.ft
.fi
.LP
To build the second database listed in 
.BR ETCDIR/slapd.conf , 
and to use three conversion sub-processes at a time,
give this command:
.LP
.nf
.ft tt
	SBINDIR/ldif2index -i ldif.input -f ETCDIR/slapd.conf -n 2 -j 3 
.ft
.fi
.LP
.SH "SEE ALSO"
.BR ldap (3),
.BR ldif (5),
.BR slapd.conf (5),
.BR ldbmcat (8),
.BR edb2ldif (8)
.LP
"The SLAPD and SLURPD Administrator's Guide"
.SH ACKNOWLEDGEMENTS
.B	OpenLDAP
is developed and maintained by The OpenLDAP Project (http://www.openldap.org/).
.B	OpenLDAP
is derived from University of Michigan LDAP 3.3 Release.  


CHANGES
OpenLDAP 2.0 Change Log

OpenLDAP 2.0.X Engineering
	Fixed -lldap TLS data ready bug (ITS#821)
	Fixed -lldap abandon unsent op bug (ITS#837)
	Fixed -lldap URL search w/o host bug (ITS#843)
	Fixed -lldap referral handling bugs (ITS#799 ITS#817)
	Fixed slapd spasswd mutex bug
	Fixed slapd ACL nameuid bug
	Fixed slapd SASL layering bug
	Fixed slapd unknown authc method bug (ITS#831)
	Fixed slapd TLSVerifyClient config bug
	Updated -lldap TLS/SASL error reporting
	Updated -llber large element handling
	Updated slapadd error reporting
	Updated slapd chroot handling (ITS#810)
	Updated slapd subschema subentry
	Added slapd numericString indexing
	Removed lint
	Build Environment
		Fixed make comment bug (ITS#811)
		Fixed $(DESTDIR) install (ITS#806)
		Added proxy check to passwd test
		Fixed slurpd tempdir perms (ITS#840)
	Documentation
		Updated ldappasswd(1)
		Updated ldap_url(3)
		Updated slapd.conf(5)
		Updated slapd(8)
		Updated slurpd(8)

OpenLDAP 2.0.6 Release
	Build Environment
		Fixed schema install (ITS#802)
		Added libtool customizations
		Re-enabled module support
	Documentation
		Update ldapsearch(1)

OpenLDAP 2.0.5 Release
	Fixed slapd/ldbm/search confusing log message (ITS#771)
	Fixed slapd/sasl no cred bug
	Fixed slapd idletimeout lock bug (ITS#767)
	Fixed slapd caseIgnoreIA5string indexing bug (ITS#786)
	Fixed slurpd TLS initialization bug (ITS#768)
	Fixed slurpd empty one shot hang (ITS#798)
	Fixed -lldap SASL ctx close bug (ITS#790)
	Fixed -lldap first/next attribute bug (ITS#778)
	Fixed -lldap (v2) rebind URL bug
	Fixed -lldap msgfree NULL bug (ITS#800)
	Fixed -lldif empty value bug
	Fixed -llber stringal len=0 bug
	Updated -llber bprint routine
	Build Environment
		Fixed LDFLAGS/-rpath usage
		Fixed IPv6 detection (ITS#669,ITS#770)
		Fixed VPATH schema install
		Updated libtool to 1.3.5
		Added Darwin support
		Disabled module support

OpenLDAP 2.0.4 Release
	Fixed clients printf/usage bugs (ITS#749)
	Fixed -lldap SASL interoperability bug
	Fixed -lldap PF_LOCAL declaration/call bugs
	Fixed slapd SASL log error (ITS#759)
	Fixed slapd spasswd support (ITS#751)
	Fixed slapd/tools fixed sasl_props bug (ITS#761)
	Updated slurpd SASL support
	Added slapd accept(2) sockopt workaround (ITS#760)
	Added slurpd TLS support
	Removed lint
	Build Environment
		Fixed --enable-spasswd (ITS#751)
	Documentation
		Fixed ldif(5) file:/// bug (ITS#750)
		Updated ldap_schema(3)

OpenLDAP 2.0.3 Release
	Fixed -lldap KBIND typo (ITS#717)
	Fixed slapd ;binary search request subtyping

OpenLDAP 2.0.2 Release
	Fixed KBIND (ITS#717)
	Fixed clients/tools -R handling (ITS#726)
	Fixed -lldap SASL got_max_ssf bug 
	Fixed -lldap TCP_NODELAY bug 
	Fixed -lldap TLS issues (ITS#733)
	Fixed -lldap_r NT threads (ITS#732)
	Fixed -llutil sa_flags never set bug
	Fixed ldappasswd -A -S crash (ITS#714)
	Fixed ldappasswd user argument usage
	Fixed slapd disallow bind_anon (ITS#721)
	Fixed slapd IPv6 issues (ITS#716)
	Fixed slapd MIT KPASSWD Compatibility (ITS#715)
	Fixed slapd time syntax routines (ITS#713)
	Fixed slapd/ldbm DB_PRIVATE not set bug (ITS#725)
	Fixed slapd/ldbm nextid reporting bug
	Fixed slapd syntaxes/mr routines (ITS#739)
	Fixed slurpd -r buffer overflow (ITS#722)
	Updated slapd rootDSE inappropriate op handling
	Added slapd syntax/mr routines
	Added slapd allow/disallow options
	Added slapd defaultSearchBase and DN verify (ITS#723)
	Removed lint
	Build Environment
		Added test000-rootdse
		Added Corba & Java Schema
		Updated MSVC projects for BDB 3.1
		Fixed Kerberos detection (ITS#717)
		Remove incompatible contribWare
	Documentation
		Fixed ldappasswd(1) usage
		Fixed ldapmodify(1) (ITS#719)
		Updated release documents (ITS#720)

OpenLDAP 2.0.1 Release
	Fixed StartTLS & ldaps:// client SDK
	Fixed slapd session close deadlock (ITS#704)
	Fixed slapd SSF ACLs
	Fixed slapd "entry" ACLs
	Updated slapd MAXARGS to 200
	Updated slapd nisNetgroupTriple validation routine (ITS#705)
	Updated slapd bootParameter validation routine (ITS#706)
	Updated slapd nameUID syntax to reuse DN routines for uniqueMember
	Updated client tool usage (ITS#710)
	Build Environment
		Fixed MSC tcp_close redefine
	Documentation
		Updated manual pages references
		Updated release documents
		Updated back-sql documents
		Updated "to do" list

OpenLDAP 2.0.0 Release
	Initial release of OpenLDAP 2.0

---
$OpenLDAP$


servers/slapd/tools/ldif2ldbm.c
#include "portable.h"

#include <stdio.h>

#include <ac/string.h>
#include <ac/ctype.h>
#include <ac/socket.h>
#include <ac/unistd.h>
#include <ac/wait.h>

#ifdef HAVE_SYS_PARAM_H
#include <sys/param.h>
#endif

#include "ldap_defaults.h"
#include "../slap.h"
#include "../back-ldbm/back-ldbm.h"
#include "ldif.h"

#define INDEXCMD		"ldif2index" EXEEXT
#define ID2ENTRYCMD		"ldif2id2entry" EXEEXT
#define ID2CHILDRENCMD		"ldif2id2children" EXEEXT

#define MAXARGS      		100

static void fork_child( char *prog, char *args[] );
static void	wait4kids( int nkidval );

static char	*tailorfile;
static char	*inputfile;
static int      maxkids = 1;
static int      nkids;

static void
usage( char *name )
{
	fprintf( stderr, "usage: %s -i inputfile [-d debuglevel] [-f configfile] [-j #jobs] [-n databasenumber] [-s sbindir]\n", name );
    exit( 1 );
}

int
main( int argc, char **argv )
{
	int		i, stop;
	char		*linep, *buf, *sbindir;
	char		*args[MAXARGS];
	char		buf2[20], buf3[20];
	char		line[BUFSIZ];
	char		cmd[MAXPATHLEN];
	int		lineno, elineno;
	int      	lmax, lcur;
	int		dbnum;
	ID		id;
	int		rc;
    int     cmdkids = 1;
	Backend		*be = NULL;
	struct ldbminfo *li;
	struct berval	bv;
	struct berval	*vals[2];
	Avlnode		*avltypes = NULL;

	ldbm_ignore_nextid_file = 1;

	sbindir = LDAP_SBINDIR;
	tailorfile = SLAPD_DEFAULT_CONFIGFILE;
	dbnum = -1;
	while ( (i = getopt( argc, argv, "d:e:s:f:i:j:n:" )) != EOF ) {
		switch ( i ) {
		case 'd':	/* turn on debugging */
			ldap_debug = atoi( optarg );
			break;

		case 's':	/* alternate sbindir (index cmd location) */
		case 'e':	/* accept -e for backwards compatibility */
			sbindir = strdup( optarg );
			break;

		case 'f':	/* specify a tailor file */
			tailorfile = strdup( optarg );
			break;

		case 'i':	/* input file */
			inputfile = strdup( optarg );
			break;

		case 'j':	/* number of parallel index procs */
			cmdkids = atoi( optarg );
			break;

		case 'n':	/* which config file db to index */
			dbnum = atoi( optarg ) - 1;
			break;

		default:
			usage( argv[0] );
			break;
		}
	}
	if ( inputfile == NULL ) {
		usage( argv[0] );
	} else {
		if ( freopen( inputfile, "r", stdin ) == NULL ) {
			perror( inputfile );
			exit( 1 );
		}
	}

	/*
	 * initialize stuff and figure out which backend we're dealing with
	 */

	rc = slap_init(SLAP_TOOL_MODE, "ldif2ldbm");
	if (rc != 0 ) {
		fprintf( stderr, "ldif2ldbm: slap_init failed!\n");
		exit(1);
	}

	read_config( tailorfile );

	if ( dbnum == -1 ) {
		for ( dbnum = 0; dbnum < nbackends; dbnum++ ) {
			if ( strcasecmp( backends[dbnum].be_type, "ldbm" )
			    == 0 ) {
				break;
			}
		}
		if ( dbnum == nbackends ) {
			fprintf( stderr, "No ldbm database found in config file\n" );
			exit( 1 );
		}
	} else if ( dbnum < 0 || dbnum > (nbackends-1) ) {
		fprintf( stderr, "Database number selected via -n is out of range\n" );
		fprintf( stderr, "Must be in the range 1 to %d (number of databases in the config file)\n", nbackends );
		exit( 1 );
	} else if ( strcasecmp( backends[dbnum].be_type, "ldbm" ) != 0 ) {
		fprintf( stderr, "Database number %d selected via -n is not an ldbm database\n", dbnum );
		exit( 1 );
	}

	slap_startup(dbnum);

	be = &backends[dbnum];

	/* disable write sync'ing */
	li = (struct ldbminfo *) be->be_private;
	li->li_dbcachewsync = 0;

	/*
	 * generate the id2entry index
	 */

	i = 0;
	sprintf( cmd, "%s/%s", sbindir, ID2ENTRYCMD );
	args[i++] = cmd;
	args[i++] = "-i";
	args[i++] = inputfile;
	args[i++] = "-f";
	args[i++] = tailorfile;
	args[i++] = "-n";
	sprintf( buf2, "%d", dbnum+1 );
	args[i++] = buf2;
	if ( ldap_debug ) {
		sprintf( buf3, "%d", ldap_debug );
		args[i++] = "-d";
		args[i++] = buf3;
	}
	args[i++] = NULL;
	fork_child( cmd, args );

	/*
	 * generate the dn2id and id2children indexes
	 */

	i = 0;
	sprintf( cmd, "%s/%s", sbindir, ID2CHILDRENCMD );
	args[i++] = cmd;
	args[i++] = "-i";
	args[i++] = inputfile;
	args[i++] = "-f";
	args[i++] = tailorfile;
	args[i++] = "-n";
	sprintf( buf2, "%d", dbnum+1 );
	args[i++] = buf2;
	if ( ldap_debug ) {
		sprintf( buf3, "%d", ldap_debug );
		args[i++] = "-d";
		args[i++] = buf3;
	}
	args[i++] = NULL;
	fork_child( cmd, args );

    maxkids = cmdkids;

	/*
	 * generate the attribute indexes
	 */

	i = 0;
	sprintf( cmd, "%s/%s", sbindir, INDEXCMD );
	args[i++] = cmd;
	args[i++] = "-i";
	args[i++] = inputfile;
	args[i++] = "-f";
	args[i++] = tailorfile;
	args[i++] = "-n";
	sprintf( buf2, "%d", dbnum+1 );
	args[i++] = buf2;
	if ( ldap_debug ) {
		sprintf( buf3, "%d", ldap_debug );
		args[i++] = "-d";
		args[i++] = buf3;
	}
	args[i++] = NULL;		/* will hold the attribute name */
	args[i++] = NULL;

	id = 0;
	stop = 0;
	buf = NULL;
	lineno = 0;
	lcur = lmax = 0;
	vals[0] = &bv;
	vals[1] = NULL;
	while ( ! stop ) {
		char		*type, *val, *s;
		ber_len_t	vlen;
		int			indexmask, syntaxmask;

		if ( fgets( line, sizeof(line), stdin ) != NULL ) {
			int     len;

			lineno++;
			len = strlen( line );
			while ( lcur + len + 1 > lmax ) {
				lmax += BUFSIZ;
				buf = (char *) ch_realloc( buf, lmax );
			}
			strcpy( buf + lcur, line );
			lcur += len;
		} else {
			stop = 1;
		}
		if ( line[0] == '\n' || stop && buf && *buf ) {
			id++;
			s = buf;
			elineno = 0;
			while ( (linep = ldif_getline( &s )) != NULL ) {
				elineno++;
				if ( ldif_parse_line( linep, &type, &val, &vlen )
				    != 0 ) {
					Debug( LDAP_DEBUG_PARSE,
			    "bad line %d in entry ending at line %d ignored\n",
					    elineno, lineno, 0 );
					continue;
				}

				if ( !isascii( *type ) || isdigit( *type ) )
					continue;

				type = strdup( type );
				if ( avl_insert( &avltypes, type, (AVL_CMP) strcasecmp,
				    avl_dup_error ) != 0 ) {
					free( type );
				} else {
					attr_masks( be->be_private, type,
					    &indexmask, &syntaxmask );
					if ( indexmask ) {
						args[i - 2] = type;
						fork_child( cmd, args );
					}
				}
			}
			*buf = '\0';
			lcur = 0;
		}
	}

	wait4kids( -1 );

	slap_shutdown(dbnum);

	slap_destroy();

	return( 0 );
}

#ifdef WIN32

static HANDLE processes[MAXIMUM_WAIT_OBJECTS];

static void
fork_child( char *prog, char *args[] )
{
    PROCESS_INFORMATION proc_info;
    PROCESS_INFORMATION *pinfo = &proc_info;
    STARTUPINFO  start_info;
    int i;
    char cmdLine[2048];

    wait4kids( maxkids );

    i = 1;
    memset( &start_info, 0, sizeof(STARTUPINFO) );
    memset( cmdLine, 0, sizeof(cmdLine) );
    strcpy( cmdLine, prog );
    while ( args[i] != NULL )
    {
        strcat( cmdLine, " " );
        strcat( cmdLine, args[i] );
        i++;
    }

    if ( !CreateProcess( NULL, cmdLine, NULL, NULL,
                         TRUE, CREATE_NEW_CONSOLE,
                         NULL, NULL, &start_info, pinfo ) )
    {
        fprintf( stderr, "Could not create %s: ", prog );
        perror( "CreateProcess" );
        exit (-1);
    }

    processes[nkids] = proc_info.hProcess;
    nkids++;

}

static void
wait4kids( int nkidval )
{
    DWORD  wait_index;
    while( nkids >= nkidval )
    {
        wait_index = WaitForMultipleObjects( nkids, processes, FALSE, INFINITE );
        /*
        *  processes[wait_index] completed.  Move any remaining indexes into its
        *  place in the array so it stays filled.
        */
        if ( nkids > 1 )
        {
            memcpy ( &processes[wait_index], &processes[wait_index+1], sizeof(HANDLE)*(nkids-1) );
            processes[nkids] = 0;
        }
        nkids--;
    }
}

#else

static void
fork_child( char *prog, char *args[] )
{
	int	pid;

	wait4kids( maxkids );

	switch ( pid = fork() ) {
	case 0:		/* child */
		execvp( prog, args );
		fprintf( stderr, "%s: ", prog );
		perror( "execv" );
		exit( -1 );
		break;

	case -1:	/* trouble */
		fprintf( stderr, "Could not fork to run %s\n", prog );
		perror( "fork" );
		break;

	default:	/* parent */
		nkids++;
		break;
	}
}

static void
wait4kids( int nkidval )
{
	int		status;
	unsigned char	*p;

	while ( nkids >= nkidval ) {
		wait( &status );
		p = (unsigned char *) &status;
		if ( p[sizeof(int) - 1] == 0177 ) {
			fprintf( stderr,
			    "stopping: child stopped with signal %d\n",
			    p[sizeof(int) - 2] );
		} else if ( p[sizeof(int) - 1] != 0 ) {
			fprintf( stderr, 
			    "stopping: child terminated with signal %d\n",
			    p[sizeof(int) - 1] );
			exit( p[sizeof(int) - 1] );
		} else if ( p[sizeof(int) - 2] != 0 ) {
			fprintf( stderr, 
			    "stopping: child exited with status %d\n",
			    p[sizeof(int) - 2] );
			exit( p[sizeof(int) - 2] );
		} else {
			nkids--;
		}
	}
}

#endif
